Add support for building ANGLE tests

kkinnunen-apple opened the pull request:
#### c3ea0df079031dd5efc2e545eb9431b452496ea8
<pre>
Add support for building ANGLE tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=293798">https://bugs.webkit.org/show_bug.cgi?id=293798</a>
<a href="https://rdar.apple.com/152301676">rdar://152301676</a>

Reviewed by Elliott Williams.

Build relevant parts of angle_end2end_tests and angle_unittests.

Adds following:
 - libutils static library for test utils
 - libANGLE static library for test compilation of ANGLE
   - links libtranslator
 - ANGLEEnd2EndTests
   - links libANGLE
   - links libutil
 - ANGLEUnitTests
   - links libANGLE
   - links libutils

libutils depends on gtest headers, so add Product Dependencies build
phase that depends on then gtest-framework.

Does not yet modify anything with libANGLE-dynamic build, the shipping
configuration.

Does not yet modify any default build, so all targets are built outside
of the main &quot;Everything up to WebKit + Tools&quot;.

Build the tests with:
   make debug SCHEME=&quot;Tools (ANGLE)&quot;

Tools (ANGLE) is an existing scheme.

Later commits will add more tests in, support for test expectations,
DEQP test runner, fix iosmac to actually run the tests.

* Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj:
* Source/ThirdParty/ANGLE/ANGLE.xcodeproj/xcshareddata/xcschemes/ANGLE (static).xcscheme:
* Source/ThirdParty/ANGLE/Configurations/ANGLE-static.xcconfig: Copied from Source/ThirdParty/ANGLE/Configurations/BaseTarget.xcconfig.
* Source/ThirdParty/ANGLE/Configurations/ANGLEEnd2EndTests.xcconfig: Added.
* Source/ThirdParty/ANGLE/Configurations/ANGLEUnitTests.xcconfig: Added.
* Source/ThirdParty/ANGLE/Configurations/BaseTarget.xcconfig:
* Source/ThirdParty/ANGLE/Configurations/utils.xcconfig: Copied from Source/ThirdParty/ANGLE/Configurations/BaseTarget.xcconfig.
* Source/ThirdParty/ANGLE/WebKit/End2EndTestsSourcesPlatform.cpp: Added.
* Source/ThirdParty/ANGLE/WebKit/TestUtilSourcesPlatform.mm: Added.
* Source/ThirdParty/ANGLE/WebKit/UtilsSourcesPlatform-mm.mm: Added.
* Source/ThirdParty/ANGLE/WebKit/UtilsSourcesPlatform.cpp: Added.
* Source/ThirdParty/ANGLE/src/common/debug.h:
* Source/ThirdParty/ANGLE/src/tests/angle_unittest_main.cpp:
(main):
* Source/ThirdParty/ANGLE/src/tests/egl_tests/EGLContextCompatibilityTest.cpp:
(RegisterContextCompatibilityTests):
* Source/ThirdParty/ANGLE/src/tests/egl_tests/EGLDisplayTest.cpp:
(TEST_P):
* Source/ThirdParty/ANGLE/src/tests/gl_tests/CopyCompressedTextureTest.cpp:
(angle::CopyCompressedTextureTest::checkExtensions const):
* Source/ThirdParty/ANGLE/src/tests/gl_tests/ImageTestMetal.mm:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/PixelLocalStorageTest.cpp:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/ProvokingVertexTest.cpp:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/RobustResourceInitTest.cpp:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/SamplersTest.cpp:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/TextureTest.cpp:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/TransformFeedbackTest.cpp:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/UniformTest.cpp:
* Source/ThirdParty/ANGLE/util/ios/ios_main.mm:
(-[AngleUtilAppDelegate application:didFinishLaunchingWithOptions:]):
(ios_main):
* Source/ThirdParty/ANGLE/util/posix/test_utils_posix.cpp:
(angle::WriteDebugMessage):

Canonical link: <a href="https://commits.webkit.org/295737@main">https://commits.webkit.org/295737@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/206aec9ef9e5ef931a7d7163f9bea11ebef8aa10

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106005 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25754 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16148 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111201 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56601 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26408 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34257 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/80522 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109009 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20837 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95668 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60843 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/20444 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13765 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56039 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90214 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13801 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114057 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33143 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/24456 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/89601 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33507 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91896 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89285 "Found 100 new API test failures: /TestWebKit:WebKit.PageLoadDidChangeLocationWithinPage, /TestWebKit:WebKit.AboutBlankLoad, /TestWebKit:WebKit.PreventEmptyUserAgent, /TestWebKit:WebKit.PrivateBrowsingPushStateNoHistoryCallback, /WebKitGTK/TestInputMethodContext:/webkit/WebKitInputMethodContext/content-type, /WebKitGTK/TestUIClient:/webkit/WebKitWebView/pointer-lock-permission-request, /TestWebKit:WebKit.HitTestResultNodeHandle, /TestWebKit:WebKit2TextFieldBeginAndEditEditingTest.TextFieldDidBeginShouldNotBeDispatchedForAlreadyFocusedField, /TestWebKit:WebKit.Find, /WebKitGTK/TestWebProcessExtensions:/webkit/WebKitWebProcessExtension/dom-input-element-is-user-edited ... (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22755 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34151 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11969 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28701 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33068 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32814 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36163 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34412 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj:
```diff
@@ -3928,6 +5465,7 @@
 		};
 		FFDA50D5269F895400AE11E2 /* Bake Metal Library to NSData */ = {
 			isa = PBXShellScriptBuildPhase;
+			alwaysOutOfDate = 1;
```
Why do we need this? It'll make the script phase run on every build, which is bad for incremental build performance. You should make the input and output paths sufficient to only re-run when necessary.

kkinnunen-apple replied:
I don't know where that comes from in Xcode. Removed this.

emw-apple commented on Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj:
We talked about this already, but you shouldn't need to embed gmock and gtest. Just add gtest as a linked framework, and Xcode will figure out the implicit dependency.

kkinnunen-apple replied:
Adding gtest-framework as a dependency *links* the gtest to libutils. This is not the intention.

Added it as a Product Dependency copy files phase.

emw-apple reviewed (changes requested):
Makes sense! I see a few things that we should fix before merging.
 
> Build the tests with:
>    make debug SCHEME="Tools (ANGLE)"

I don't see a `Tools (ANGLE).xcscheme` in the diff, do you need to mark it as "shared" and check it in? Also, if these targets are only built outside of the main "Everything up to WebKit + Tools" scheme, CI won't check that the build. Is that okay?

emw-apple commented on Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj:
```diff
@@ -4816,6 +6719,54 @@
+			isa = XCBuildConfiguration;
+			baseConfigurationReference = 7BFAF8332DE6EBA700327F7F /* ANGLEUnitTests.xcconfig */;
+			buildSettings = {
+				PRODUCT_NAME = "$(TARGET_NAME)";
```
Xcode adds this build setting in the pbxproj by default, but we don't need it because we set PRODUCT_NAME from xcconfig. Could you remove this override and the ones below?

kkinnunen-apple committed:
Add support for building ANGLE tests
https://bugs.webkit.org/show_bug.cgi?id=293798
rdar://152301676

Reviewed by Elliott Williams.

Build relevant parts of angle_end2end_tests and angle_unittests.

Adds following:
 - libutils static library for test utils
 - libANGLE static library for test compilation of ANGLE
   - links libtranslator
 - ANGLEEnd2EndTests
   - links libANGLE
   - links libutil
 - ANGLEUnitTests
   - links libANGLE
   - links libutils

libutils depends on gtest headers, so add Product Dependencies build
phase that depends on then gtest-framework.

Does not yet modify anything with libANGLE-dynamic build, the shipping
configuration.

Does not yet modify any default build, so all targets are built outside
of the main "Everything up to WebKit + Tools".

Build the tests with:
   make debug SCHEME="Tools (ANGLE)"

Tools (ANGLE) is an existing scheme.

Later commits will add more tests in, support for test expectations,
DEQP test runner, fix iosmac to actually run the tests.

* Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj:
* Source/ThirdParty/ANGLE/ANGLE.xcodeproj/xcshareddata/xcschemes/ANGLE (static).xcscheme:
* Source/ThirdParty/ANGLE/Configurations/ANGLE-static.xcconfig: Copied from Source/ThirdParty/ANGLE/Configurations/BaseTarget.xcconfig.
* Source/ThirdParty/ANGLE/Configurations/ANGLEEnd2EndTests.xcconfig: Added.
* Source/ThirdParty/ANGLE/Configurations/ANGLEUnitTests.xcconfig: Added.
* Source/ThirdParty/ANGLE/Configurations/BaseTarget.xcconfig:
* Source/ThirdParty/ANGLE/Configurations/utils.xcconfig: Copied from Source/ThirdParty/ANGLE/Configurations/BaseTarget.xcconfig.
* Source/ThirdParty/ANGLE/WebKit/End2EndTestsSourcesPlatform.cpp: Added.
* Source/ThirdParty/ANGLE/WebKit/TestUtilSourcesPlatform.mm: Added.
* Source/ThirdParty/ANGLE/WebKit/UtilsSourcesPlatform-mm.mm: Added.
* Source/ThirdParty/ANGLE/WebKit/UtilsSourcesPlatform.cpp: Added.
* Source/ThirdParty/ANGLE/src/common/debug.h:
* Source/ThirdParty/ANGLE/src/tests/angle_unittest_main.cpp:
(main):
* Source/ThirdParty/ANGLE/src/tests/egl_tests/EGLContextCompatibilityTest.cpp:
(RegisterContextCompatibilityTests):
* Source/ThirdParty/ANGLE/src/tests/egl_tests/EGLDisplayTest.cpp:
(TEST_P):
* Source/ThirdParty/ANGLE/src/tests/gl_tests/CopyCompressedTextureTest.cpp:
(angle::CopyCompressedTextureTest::checkExtensions const):
* Source/ThirdParty/ANGLE/src/tests/gl_tests/ImageTestMetal.mm:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/PixelLocalStorageTest.cpp:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/ProvokingVertexTest.cpp:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/RobustResourceInitTest.cpp:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/SamplersTest.cpp:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/TextureTest.cpp:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/TransformFeedbackTest.cpp:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/UniformTest.cpp:
* Source/ThirdParty/ANGLE/util/ios/ios_main.mm:
(-[AngleUtilAppDelegate application:didFinishLaunchingWithOptions:]):
(ios_main):
* Source/ThirdParty/ANGLE/util/posix/test_utils_posix.cpp:
(angle::WriteDebugMessage):

Canonical link: https://commits.webkit.org/295737@main
